### PR TITLE
[NNAPI EP] Add infrastructure for Op Reduction

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: [NNAPI EP] Add infrastructure for Op Reduction

**Motivation and Context**
- We want to not link the code we are not using (if certain ops are not used for the build) to reduce binary size. Adding infrastructure for that.
- The build script is not updated to handle reducing ops yet, that will be another PR for this

### CODE REVIEW

**Comment by skottmckay** [comment]
>   return op_map;

Would it make the code clearer to simply have a struct with  std::vector<unique_ptr<IOpBuilder>> providing storage, and an unordered_map<std::string, IOpBuilder*> for the lookup? CreateSharedOpBuilder could add an entry to the vector, and entries to the map for all the ops covered by the op builder on the first call so the map is fully initialized upfront instead of bit by bit via FindOpBuilder. 

---
Refers to: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:2289 in ba44f4a. [](commit_id = ba44f4a407c3ba1dbe0817aa2cad89964d252d77, deletion_comment = False)

File: onnxruntime/test/providers/nnapi/nnapi_basic_test.cc [hunk 1]
```diff
@@ -67,6 +72,20 @@ TEST(NnapiExecutionProviderTest, ReshapeFlattenTest) {
 #endif
 }
 
+#if defined(__ANDROID__)
+// This is to verify the op_builders and op_support_checkers are consistent
+TEST(NnapiExecutionProviderTest, CreateOpBuilderAndOpSupportCheckerTest) {
+  const auto& op_builders = nnapi::GetOpBuilders();
```
Comment (skottmckay) [inline onnxruntime/test/providers/nnapi/nnapi_basic_test.cc:78]: If the checkers and builders are 1:1 would it make sense to combine the them into one instance so you have an 'op handler' (or something like that) which is composed of the checking and building capabilities. 

Would halve the registration logic and help ensure things are consistent.

Comment (gwang-msft) [reply onnxruntime/test/providers/nnapi/nnapi_basic_test.cc:78]: The thing is OpBuilder cannot build for non-android platform, so I have to separate them, this is not ideal, I plan to have some more refactoring after 1.6 to make them more consistent

**Comment by gwang-msft** [comment]
>   return op_map;

I have thought about this, for example, move all the Abs/Exp/Floor ... into a single AddUnaryOpBuilder, and only call it once, however, this will require I add logic to tell if an op is an unaryOp into some other pieces of code, such as the reduction python script and modelbuilder, which will all require to be updated if I add a new unary op, I would prefer easier maintenance than a tiny perf hit here, :smile: 
I will revisit this when I try to refactor the code after 1.6


---
In reply to: [729452180](https://github.com/microsoft/onnxruntime/pull/5841#issuecomment-729452180) [](ancestors = 729452180)

---
Refers to: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:2289 in ba44f4a. [](commit_id = ba44f4a407c3ba1dbe0817aa2cad89964d252d77, deletion_comment = False)

**Comment by skottmckay** [comment]
>   return op_map;

Wasn't perf I was thinking about - more that the logic in CreateSharedOpBuilder was somewhat confusing. So it was about simplifying that a little, avoiding the shared_ptr, but not changing the structure much. 

e.g. 
struct OpBuilderRegistrations {
  std::vector<std::unique_ptr<IOpBuilder>> builders;
  unordered_map<std::string, IOpBuilder*> op_builder_map;
}

void BinaryOpBuilder::CreateSharedOpBuilder( const std::string& op_type, OpBuilderRegistrations & op_registrations) {
  const sstd::vector<std::string> current_ops = {"Add", "Sub", "Mul", "Div", "QLinearAdd"};
  if (op_registrations.op_builder_map.find(current_ops.front()) == op_registrations.op_builder_map.cend()){
    op_registrations.builders.push_back(onnxruntime::make_unique<BinaryOpBuilder>());
    // add entries for all ops covered by this builder
    for (const auto& op : current_ops)
      op_registrations.op_builder_map.insert({op, op_registrations.builders.back().get());
  }
}

Possibly there's a way to templatize that given you're really just plugging in a list of ops and a builder type, and the rest is generic (unless the builder ctor requires args).

Just something to consider.

---
In reply to: [729471069](https://github.com/microsoft/onnxruntime/pull/5841#issuecomment-729471069) [](ancestors = 729471069,729452180)

---
Refers to: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:2289 in ba44f4a. [](commit_id = ba44f4a407c3ba1dbe0817aa2cad89964d252d77, deletion_comment = False)

File: onnxruntime/test/providers/nnapi/nnapi_basic_test.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/test/providers/nnapi/nnapi_basic_test.cc:78]: OK. The macro to create the builder could just return an instance of IOpBuilder that threw if used couldn't it? That may confine the #ifdef on platform just to a few macros.

---
In reply to: [525842700](https://github.com/microsoft/onnxruntime/pull/5841#discussion_r525842700) [](ancestors = 525842700)

**Review by skottmckay — APPROVED** [review]
:shipit:

**Review by skottmckay — APPROVED** [review]

**Pull request merged**